pkg/maintenance/alertmanager.go

```diff
@@ -0,0 +1,173 @@
+	"github.com/go-openapi/strfmt"
+	routev1 "github.com/openshift/api/route/v1"
+	amSilence "github.com/prometheus/alertmanager/api/v2/client/silence"
+	amv2Models "github.com/prometheus/alertmanager/api/v2/models"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+	"time"
+)
+
+var (
+	alertManagerNamespace          = "openshift-monitoring"
+	alertManagerRouteName          = "alertmanager-main"
+	alertManagerServiceAccountName = "prometheus-k8s"
+	alertManagerBasePath           = "/api/v2/"
+)
+
+type alertManagerMaintenance struct {
+	client    client.Client
+	transport *httptransport.Runtime
+}
+
+func newAlertManagerMaintenance(client client.Client) (Maintenance, error) {
+	transport, err := getTransport(client)
+	if err != nil {
+		return nil, err
+	}
+
+	transport.DefaultAuthentication, err = getAuthentication(client)
+	if err != nil {
+		return nil, err
+	}
+
+	return &alertManagerMaintenance{
+		client:    client,
+		transport: transport,
+	}, nil
+}
+
+func getTransport(c client.Client) (*httptransport.Runtime, error) {
+	amRoute := &routev1.Route{}
+	err := c.Get(
+		context.TODO(),
+		types.NamespacedName{Namespace: alertManagerNamespace, Name: alertManagerRouteName},
+		amRoute,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return httptransport.New(
+		amRoute.Spec.Host,
+		alertManagerBasePath,
+		[]string{"https"},
+	), nil
+}
+
+func getAuthentication(c client.Client) (runtime.ClientAuthInfoWriter, error) {
+	sl := &corev1.SecretList{}
+	err := c.List(
+		context.TODO(),
+		sl,
+		&client.ListOptions{Namespace: alertManagerNamespace},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var token string
+	for _, s := range sl.Items {
+		if strings.Contains(s.Name, alertManagerServiceAccountName+"-token") {
+			token = string(s.Data["token"])
+		}
+	}
+
+	return httptransport.BearerToken(token), nil
+}
+
+func (amm *alertManagerMaintenance) Start() error {
+	createdBy := "managed-upgrade-operator"
+	comment := "Test comment"
+	startsAt := strfmt.DateTime(time.Now().UTC())
+	endsAt := strfmt.DateTime(time.Now().Add(time.Minute * 5))
+
+	name := "alertname"
+	isRegex := true
+	value := "[A-Z].*"
+	psParams := &amSilence.PostSilencesParams{
+		Silence: &amv2Models.PostableSilence{
+			Silence: amv2Models.Silence{
+				CreatedBy: &createdBy,
+				Comment:   &comment,
+				EndsAt:    &endsAt,
+				StartsAt:  &startsAt,
+				Matchers: amv2Models.Matchers{
+					{
+						Name:    &name,
+						IsRegex: &isRegex,
+						Value:   &value,
+					},
+				},
+			},
+		},
+		Context:    context.TODO(),
+		HTTPClient: &http.Client{},
+	}
+
+	silenceClient := amSilence.New(amm.transport, strfmt.Default)
+	_, err := silenceClient.PostSilences(psParams)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (amm *alertManagerMaintenance) End() error {
+	dParams := &amSilence.DeleteSilenceParams{
+		SilenceID:  "",
```

Comment:
Just curious (I don't know the answer) - does having no SilenceIDs specified result in deleting all silences, or will End() need to query for the existing silence IDs first in order to populate this?

Reply:
The api needs an actual id for each delete.
Once I start integrating it ill see do we need to store an id or maybe just do a get and check `CreatedBy` = "managed-upgrade-operator" or whatever.